### PR TITLE
Feature/Chainlink Integrations

### PR DESCRIPTION
**Pull request by HenryHengZJ**
## What does this PR do?
Chainlink integration
Docs: https://docs.chain.link

## Tasks Completed
- [x] Add ChainLink action node. ChainLink action node should have some key APIs from the docs implemented such as price feeds, nft floor price.

## Pending Tasks To Be Completed
- [ ] Add VRF Randomness

## Related PRs and Issues
The Pull Request is related to this https://github.com/Outerbridgeio/Outerbridge/issues/87

## (Optional) Screenshot
image

### CODE REVIEW

File: packages/components/nodes/ChainLink/ChainLink.ts [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+		} catch(e) {
+			throw handleErrorMessage(e);
+		}
+	}
+}
+
+const getInputParametersData = async(network: NETWORK, dataType: string) => {
+
+    const returnData = [] as INodeOptionsValue[];
```
Comment (tylim88) [inline packages/components/nodes/ChainLink/ChainLink.ts:235]: prefer type annotation over type cast

 `const returnData: INodeOptionsValue[] = []`

File: packages/components/nodes/ChainLink/ChainLink.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+			const provider = await getNetworkProvider(
+				networksData.networkProvider as NETWORK_PROVIDER,
+				network,
+				credentials,
+				networksData.jsonRPC as string,
+				networksData.websocketRPC as string,
+			)
+
+			if (!provider) throw new Error('Invalid Network Provider');
```
Comment (tylim88) [inline packages/components/nodes/ChainLink/ChainLink.ts:202]: though in this case it doesnt matter much but short hand if statement potentially make debugging difficult

```ts
if(condition)   
doSomething() // may run depends on condition
```

one day you want to trouble shoot

```ts
if(condition) 
// add console.log for debugging  
console.log("something") // may run depends on condition
doSomething() // will run regardless of condition`
```

File: packages/components/nodes/ChainLink/supportedNetwork.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    [NETWORK.AVALANCHE]: 'Avalanche Mainnet',
+    [NETWORK.AVALANCHE_TESTNET]: 'Avalanche Testnet',
+    [NETWORK.SOLANA]: 'Solana Mainnet',
+    [NETWORK.SOLANA_DEVNET]: 'Solana Devnet',
+    [NETWORK.HARMONY]: 'Harmony Mainnet',
+    [NETWORK.MOONRIVER]: 'Moonriver Mainnet',
+    [NETWORK.MOONBEAM]: 'Moonbeam Mainnet',
+    [NETWORK.METIS]: 'Metis Mainnet',
+    [NETWORK.KLATYN_TESTNET]: 'Klaytn Baobab testnet',
+} as any;
```
Comment (tylim88) [inline packages/components/nodes/ChainLink/supportedNetwork.ts:28]: prefer narrowly typed const assertion `as const`

if narrowed type is not compatible with other type for now, prefer type cast as index signature `as Record<string, string>`

File: packages/components/nodes/ChainLink/supportedNetwork.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    },
+    {
+        label: NETWORK_LABEL.KLATYN_TESTNET,
+        name: NETWORK.KLATYN_TESTNET,
+        parentGroup: 'Klaytn Baobab testnet',
+        hide: {
+            'actions.operation': ['getProofReserve', 'getNFTFloorPricing']
+        }
+    },
+] as INodeOptionsValue[];
```
Comment (tylim88) [inline packages/components/nodes/ChainLink/supportedNetwork.ts:215]: prefer type annotation over type cast

File: packages/components/src/ChainNetwork.ts [hunk 5]
```diff
@@ -652,8 +829,66 @@ export const polygonMumbaiRPC = [
 ] as string[];
```
Comment (tylim88) [inline packages/components/src/ChainNetwork.ts:829]: type annotation over type cast

File: packages/components/nodes/ChainLink/ChainLink.ts (hunk 2 above)
Comment (HenryHengZJ) [reply packages/components/nodes/ChainLink/ChainLink.ts:202]: yeah I had it this way cus its cleaner. What do you suggest?
```
if(condition) {
  doSomething() // may run depends on condition
}
```

File: packages/components/nodes/ChainLink/ChainLink.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+                    },
+                    {
+                        label: 'Get NFT Floor Pricing',
+                        name: 'getNFTFloorPricing',
+                        description: 'Get the lowest price of an NFT in a collection using ChainLink Oracle, ONLY available on Goerli'
+                    }
+                ],
+				default: 'getPriceFeeds'
+			},
+		] as INodeParams[];
```
[the remark this answers is not in the document]
Comment (tylim88) [reply packages/components/nodes/ChainLink/ChainLink.ts]: This is good, it has scope bracket

File: packages/components/nodes/ChainLink/ChainLink.ts (hunk 2 above)
Comment (tylim88) [reply packages/components/nodes/ChainLink/ChainLink.ts:202]: This is good, it has scope bracket

File: packages/components/nodes/ChainLink/ChainLink.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tylim88) [reply packages/components/nodes/ChainLink/ChainLink.ts]: You should remove it

The old code can leave them as it is now, don't spend too much time on old code

But in the future, avoid type cast if possible

[the remark this answers is not in the document]
Comment (tylim88) [reply packages/components/nodes/ChainLink/ChainLink.ts]: further explanation why type cast should be avoided
image

https://www.typescriptlang.org/play?#code/C4TwDgpgBAggQlAvAbygQwFwGdgCcCWAdgOYA0UARhoQK4C2FEuUAvgFBsDGA9oTukiipMAIgAeI1uiywEAejlQRFSfhl01WIsUo1gUQtyihIUJrm64O1jjz76q8QcIxKJUhcfDRzl8stV1TW0AQiA

**Comment by tylim88** [comment]
I think we don't need to spend too much time on type

focus on getting features first

**Review by tylim88 — DISMISSED** [review]

File: packages/components/src/ChainNetwork.ts (hunk 5 above)
Comment (HenryHengZJ) [reply packages/components/src/ChainNetwork.ts:829]: yeah we have alot of type cast, so could be a seperate feature issue to work on those

**Review by tylim88 — APPROVED** [review]

**Pull request merged**